OpenRA.Mods.Common/Activities/Resupply.cs

```diff
@@ -17,48 +17,96 @@
 
 namespace OpenRA.Mods.Common.Activities
 {
-	public class Repair : Activity
+	public class Resupply : Activity
 	{
 		readonly IHealth health;
 		readonly RepairsUnits[] allRepairsUnits;
 		readonly Target host;
 		readonly WDist closeEnough;
 		readonly Repairable repairable;
+		readonly RepairableNear repairableNear;
+		readonly Rearmable rearmable;
+		readonly INotifyResupply[] notifyResupplies;
 
 		int remainingTicks;
-		bool played = false;
+		bool played;
+		bool paused;
+		bool repairComplete;
+		bool rearmComplete;
+		ResupplyType activeResupplyTypes = ResupplyType.None;
 
-		public Repair(Actor self, Actor host, WDist closeEnough)
+		public Resupply(Actor self, Actor host, WDist closeEnough)
 		{
 			this.host = Target.FromActor(host);
 			this.closeEnough = closeEnough;
 			allRepairsUnits = host.TraitsImplementing<RepairsUnits>().ToArray();
 			health = self.TraitOrDefault<IHealth>();
 			repairable = self.TraitOrDefault<Repairable>();
+			repairableNear = self.TraitOrDefault<RepairableNear>();
+			rearmable = self.TraitOrDefault<Rearmable>();
+			notifyResupplies = host.TraitsImplementing<INotifyResupply>().ToArray();
+
+			repairComplete = health == null || health.DamageState == DamageState.Undamaged
+				|| !allRepairsUnits.Any()
+				|| ((repairable == null || !repairable.Info.RepairActors.Contains(host.Info.Name))
+					&& (repairableNear == null || !repairableNear.Info.RepairActors.Contains(host.Info.Name)));
+
+			if (!repairComplete)
+				activeResupplyTypes |= ResupplyType.Repair;
+
+			rearmComplete = rearmable == null || !rearmable.Info.RearmActors.Contains(host.Info.Name) || rearmable.RearmableAmmoPools.All(p => p.FullAmmo());
+			if (!rearmComplete)
+				activeResupplyTypes |= ResupplyType.Rearm;
 		}
 
 		protected override void OnFirstRun(Actor self)
 		{
-			if (host.Actor.IsDead)
+			if (host.Type == TargetType.Invalid)
 				return;
 
-			foreach (var depot in host.Actor.TraitsImplementing<INotifyRepair>())
-				depot.BeforeRepair(host.Actor, self);
+			if (!repairComplete || !rearmComplete)
+				foreach (var notifyResupply in notifyResupplies)
+					notifyResupply.BeforeResupply(host.Actor, self, activeResupplyTypes);
+
+			// Reset the ReloadDelay to avoid any issues with early cancellation
+			// from previous reload attempts (explicit order, host building died, etc).
+			// HACK: this really shouldn't be managed from here
+			if (!rearmComplete)
+				foreach (var pool in rearmable.RearmableAmmoPools)
+					pool.RemainingTicks = pool.Info.ReloadDelay;
 		}
 
 		public override Activity Tick(Actor self)
 		{
 			if (IsCanceling)
-			{
-				if (remainingTicks-- == 0)
-					return NextActivity;
+				return NextActivity;
 
-				return this;
-			}
+			if (host.Type == TargetType.Invalid || health == null)
```

Comment:
What is the purpose of this health check? Do we really want to prevent units without a `Health` trait from rearming?

Reply:
Just a left-over I overlooked.